### PR TITLE
result aggregation in resume API

### PR DESCRIPTION
**Pull request by kaiserd**
This PR addresses https://github.com/status-im/nim-waku/issues/809

- updates `resume` and `queryLoop` implementing the logic described in https://github.com/status-im/nim-waku/issues/809
- updates the existing resume test and adds a test for resuming from an empty `peerList`
- updates the [changelog](https://github.com/status-im/nim-waku/blob/master/CHANGELOG.md)
- updates the waku v2 API [description](https://github.com/status-im/nim-waku/blob/master/docs/api/v2/node.md)

The code for awaiting the futures in `resume` is clunky. I could not find a version of `allFutures` that returns a `Future[seq[T]]`.
What I ended up using is `allFinished*[T](futs: varargs[Future[T]]): Future[seq[Future[T]]]` which requires to go through the already finshed futures again.
At least, this allows the futures to be awaited in parallel.
Did I overlook something?

Another point that came up:
Currently, if no peer in the peerList answers, resume exits with an error.
It could instead ask known peers, which is the case when peerList is empty.
What do you think @staheri14 @jm-clius ?

### CODE REVIEW

File: waku/v2/protocol/waku_store/waku_store.nim
Comment (jm-clius) [inline waku/v2/protocol/waku_store/waku_store.nim:621]: Any advantage to doing this rather than just use `await allFutures()`? Seems like that will at least avoid double-assigning to `futureList`? It will still be necessary to retrieve the completed values iteratively, though, as you did below.

File: waku/v2/protocol/waku_store/waku_store.nim
Comment (jm-clius) [inline waku/v2/protocol/waku_store/waku_store.nim:625]: Since we "know" that these futures have all completed, why not rather just `read` the value (with a sanity check for completeness too). I think this will improve readability.
For example:
```
if fut.completed():
  let successResult = fut.read()
```
or something similar should be safe to do.

File: waku/v2/protocol/waku_store/waku_store.nim
Comment (jm-clius) [inline waku/v2/protocol/waku_store/waku_store.nim:626]: I think it may be possible to map `futureList` to `messagesList` using the [`mapIt()`](https://nim-lang.org/docs/sequtils.html#mapIt.t%2Ctyped%2Cuntyped) template and `concat`ting the resulting sequence-of-sequences. Compressing too much logic into a single line is probably not a good idea, so feel free to ignore this suggestion, or just applying it to some steps. :)

**Review by jm-clius — APPROVED** [review]
This LGTM! Thanks. I've left some comments about possible ways to make the code more readable, but feel free to apply only where you deem necessary/helpful.

File: waku/v2/protocol/waku_store/waku_store.nim
Comment (jm-clius) [reply waku/v2/protocol/waku_store/waku_store.nim:625]: You could also encompass this logic into the `op proc` parameter of a `mapIt` function and directly map your `seq[Future[MessagesResult]]` to a `seq[MessagesResult]`. With a bit more logic, and concatenating the different responses, the original `seq` of futures could also be mapped to a `seq[WakuMessage]` using a single `mapIt` function.

**Comment by jm-clius** [comment]
> The code for awaiting the futures in resume is clunky. I could not find a version of allFutures that returns a Future[seq[T]].

Seems like this did exist, but was deprecated for a number of reasons: https://github.com/status-im/nim-chronos/blob/7dc58d42b6905a7fd7531875fa76060f8f744e4e/chronos/asyncfutures2.nim#L689

You could use the `mapIt` function for sequences to map to your desired `seq`. See my comment elsewhere on the PR.

> Currently, if no peer in the peerList answers, resume exits with an error.

I think this is fine. We could decide on a future "enhancement" where we use an existing peer, but IMO if someone calls the API with an explicit list of peers it's better to return an error than transparently contacting a peer that was _not_ provided.

File: waku/v2/protocol/waku_store/waku_store.nim
Comment (kaiserd) [reply waku/v2/protocol/waku_store/waku_store.nim:621]: Thank you. I will apply this change. As I am very new to Nim, I did not know better 😅. At first I thought I could save the following iteration by directly getting a list with the results of all successfully resolved futures, and then replace the allFInished call with the respective call. But as you pointed out, using allFinished, I still have to iterate.

File: waku/v2/protocol/waku_store/waku_store.nim
Comment (kaiserd) [reply waku/v2/protocol/waku_store/waku_store.nim:625]: Thank you. I did not know about `read()`. I will apply that.

File: waku/v2/protocol/waku_store/waku_store.nim
Comment (kaiserd) [reply waku/v2/protocol/waku_store/waku_store.nim:626]: Thank you. I was looking for this, but then decided to take syntactically simple approach due to lack of Nim knowledge.
I will look into it an apply the change you suggested.

**Comment by kaiserd** [comment]
Thank you for the feedback.
In pushed commits addressing your feedback.
The second commit addresses the functional style of constructing the `messagesList`.
In Rust, there are combinators like filterMap, which would allow making this more concise. I will check later, if Nim offers something similar.

File: waku/v2/protocol/waku_store/waku_store.nim
Comment (jm-clius) [inline waku/v2/protocol/waku_store/waku_store.nim:628]: I like how explicit and readable this is, but it may be worth considering that each `mapIt` and `filterIt` returns a new `seq` which for very large sequences could be inefficient. Not really a problem here, as the `futureList` will be small.
The way to fix that would be to call [a `map` proc](https://nim-lang.org/docs/sequtils.html#map%2CopenArray%5BT%5D%2Cproc%28T%29) only once, but with a slightly more advanced logic. Something like:
```
proc mapToWakuMessages(in: Future[MessagesResult]): seq[WakuMessages] =
  if in.completed() and in.read().isOk():
    in.read().value
  else:
    @[]

let messagesList = futureList.map(mapToWakuMessages).concat()
```
(Might be some errors here, but the rough idea is to minimise seq copies)

**Comment by kaiserd** [comment]
I updated to a single map proc. However, not giving the proc directly as a parameter raises the following error:
```
Error: can raise an unlisted exception: CatchableError
```
So, for now, I defined the proc inline. (Using the `=>` sugar syntax would be more readable, imho.)
I will check, what caused that error.

**Comment by kaiserd** [comment]
Thanks @jm-clius for the tip to annotate the exception.
With that, we could also use this code if perferred:

```nim
  proc mapToWakuMessages(it: Future[MessagesResult]): seq[WakuMessage] {.raises: [CatchableError, Defect].}=
    if it.completed() and it.read().isOk():
      it.read().value
    else:
      @[]

  let messagesList = futureList.map(mapToWakuMessages).concat()
```

**Comment by staheri14** [comment]
> It could instead ask known peers, which is the case when peerList is empty.

File: tests/v2/test_waku_store.nim
Comment (staheri14) [inline tests/v2/test_waku_store.nim:718]: `empty list of candidate peers` or `offline peers`?

File: tests/v2/test_waku_store.nim
Comment (staheri14) [inline tests/v2/test_waku_store.nim:742]: Wondering why 14? (ok, I found it :), `listenSweitch2` has 4 waku messages that `listenSwitch` does not have, though it would be good to add a comment and explain that, would be helpful for debugging)

File: tests/v2/test_waku_store.nim
Comment (staheri14) [inline tests/v2/test_waku_store.nim:739]: Wondering why did you add `listenSwitch` twice? if you have a specific test scanrio, please add a small comment

**Review by staheri14 — APPROVED** [review]
Congratulations @kaiserd for your first nim-waku PR!
I have left a few comments. Going to approve the PR, but please address them before merging :)

File: CHANGELOG.md
Comment (oskarth) [inline CHANGELOG.md:14]:
```suggestion
- Resuming history via `resume` now takes the answers of all peers in `peerList` and consolidates them into one deduplicated list
```

**Review by oskarth — APPROVED** [review]
LGTM, nothing to add to above. Congrats on your first PR!

**Comment by kaiserd** [comment]
@staheri14 thank you for the feedback. I addressed your feedback and added comments.

> Can you please elaborate on this?
> 
> > It could instead ask known peers, which is the case when peerList is empty.

I thought it might make sense to ask a random peer like here
https://github.com/status-im/nim-waku/blob/024128b39e77935268e049769ffbe4a6987bf5aa/waku/v2/protocol/waku_store/waku_store.nim#L723
in case no `peerList` is provided to `resume`.
But I agree with @jm-clius   comment above, that this behavior might not be expected.
So I would just leave it as it is.

**Comment by kaiserd** [comment]
@jm-clius @staheri14 is your feedback sufficiently addressed? Just want to make sure before merging :).
@jm-clius Should I keep the current version of building `messagesList?

**Comment by jm-clius** [comment]
Yeah, you can squash and merge - I notice not all commits are signed, which you'll need to do beforehand, but afaict you addressed all comments from all reviewers, CI is passing (as well as it does :) ) and the `messagesList` building makes sense to me. 😄

**Comment by kaiserd** [comment]
Yes. The signatures were lost during a rebase :(. I reapplied them.

**Comment by kaiserd** [comment]
I rebased again on the v0.7 commit.
@jm-clius Should this PR be part of v0.7. If not, I have to update the changelog.

**Comment by jm-clius** [comment]
No, I think it's OK to merge after the release

**Comment by kaiserd** [comment]
OK. I am waiting for @staheri14 to check if I applied her feedback as she expected and then merge :).

**Comment by staheri14** [comment]
> OK. I am waiting for @staheri14 to check if I applied her feedback as she expected and then merge :).

@kaiserd LGTM!

**Pull request merged**
